peer/record.go

```diff
@@ -0,0 +1,188 @@
+package peer
+
+import (
+	"errors"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/libp2p/go-libp2p-core/crypto"
+	pb "github.com/libp2p/go-libp2p-core/peer/pb"
+	"github.com/libp2p/go-libp2p-core/record"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+// The domain string used for peer records contained in a SignedEnvelope.
+const PeerRecordEnvelopeDomain = "libp2p-peer-record"
+
+// The type hint used to identify peer records in a SignedEnvelope.
+// TODO: register multicodec
+var PeerRecordEnvelopePayloadType = []byte("/libp2p/peer-record")
+
+// ErrPeerIdMismatch is returned when attempting to sign a PeerRecord using a key that
+// does not match the PeerID contained in the record.
+var ErrPeerIdMismatch = errors.New("unable to sign peer record: key does not match record.PeerID")
+
+// PeerRecord contains information that is broadly useful to share with other peers,
+// either through a direct exchange (as in the libp2p identify protocol), or through
+// a Peer Routing provider, such as a DHT.
+//
+// Currently, a PeerRecord contains the public listen addresses for a peer, but this
+// is expected to expand to include other information in the future.
+//
+// PeerRecords are intended to be signed by the peer they describe, and there are no
+// public methods for marshalling and unmarshalling unsigned PeerRecords.
+//
+// To share a PeerRecord, first call Sign to wrap the record in a SignedEnvelope
+// and sign it with the local peer's private key:
+//
+//     rec := NewPeerRecord(myPeerId, myAddrs)
+//     envelope, err := rec.Sign(myPrivateKey)
```

Comment:
Aha, that's interesting. If `Sign()` is a member of the record type, wouldn't it have to be duplicated everywhere? A more ergonomic API is to hoist the `Sign` method to the `record` package, and call it something like `Seal`, which returns an envelope. Really, signing a record is one operation that is part of the larger act of authenticating and securing a record.

Reply:
sorry, I missed this comment earlier. I think you're right that it makes more sense as a function, now that we have a `Record` interface. 👍